### PR TITLE
[INFINITY-1466] test frameworks on overlay network, dns

### PR DESCRIPTION
**Pull request by ArtRand**
This PR only contains the code to do the following: 
0. Remove old "calico-specific" (guessing here) code for data structures that we're not going to support (e.g. netgroups).
1. Specify that a pod should join a virtual network (usually the 'dcos' overlay network) in YAML, checks that the network specified is "supported", warns if not.
2. When a pod joins the overlay, its port resources are not added to its offer requirements.
3. Confirms that the user doesn't specify container:port to host:port mappings on a virtual network that doesn't support it. 
4. Does not allow a service to migrate (upgrade, update) from a configuration where it uses does not use ports on the host machine (on overlay) to a configuration where it does (or vice versa). **NB** this is only unit-tested, not CI tested. 

This PR only contains code, updates to docs will be in another PR. This PR also does not contain changes to the /endpoints api, which will also be in another PR.

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import javafx.util.Pair;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Validates that pods do not move from a virtual network that does not use port mapping (and thus ignores the ports
+ *  as Mesos resources) to a network (virtual or Host) that does use the ports on the host machine.
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java]: double space at start

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                        String.format("New config has pod %s moving networks from %s to %s, changing it's " +
+                                        "host ports requirements from %s to %s, not allowed.",
+                                newPod.getType(), oldPod.getNetworks().toString(), newPod.getNetworks().toString(),
+                                podSpecUsesHostPorts(oldPod), podSpecUsesHostPorts(newPod))));
+            }
+        }
+        return errors;
+    }
+
+    private boolean podSpecUsesHostPorts(PodSpec podSpec) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java]: looks like this could be `static`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        }
+        for (NetworkSpec networkSpec : podSpec.getNetworks()) {
+            if (DcosConstants.networkSupportsPortMapping(networkSpec.getName())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private String podSpecOverlayNetworkIsCorrect(PodSpec podSpec) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java]: - Could be `static`?
- Maybe just return a `List<String>`
- Could be named something like `getNetworksWithInvalidPortMappings()`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            // first we check that the new pod's networks (if present) do not erroneously use port mapping
+            String offendingNetworkNames = podSpecOverlayNetworkIsCorrect(newPod);
+            if (!Strings.isNullOrEmpty(offendingNetworkNames)) {
+                errors.add(ConfigValidationError.transitionError(
+                        String.format("PodSpec[name:%s]", newPod.getType()),
+                        "null",
+                        String.format("%s", newPod.getNetworks().toString()),
+                        String.format("New config has pod %s that uses network(s) %s, port mapping is " +
+                                        "not supported on these networks. NetworkSpec: %s", newPod.getType(),
+                                offendingNetworkNames, newPod.getNetworks().toString())));
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java]: Once the names is a `List<String>`, this could use (`apache.commons.lang3`) `StringUtils.join(list, ", ")` for this log message

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            }
+            // now we check that the none of the new pods move from not using host ports to using them (or vice
+            // versa)
+            PodSpec oldPod = oldPods.get(kv.getKey());
+            if (oldPod != null && podSpecUsesHostPorts(oldPod) != podSpecUsesHostPorts(newPod)) {
+                errors.add(ConfigValidationError.transitionError(
+                        String.format("PodSpec[name:%s]", oldPod.getType()),
+                        String.format("%s", oldPod.getNetworks().toString()),
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java]: Could strip out the `String.format()` and just have `[old|new]Pod.getNetworks().toString()`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/Capabilities.java [hunk 6]
```diff
@@ -32,9 +32,9 @@ public boolean supportsGpuResource() throws IOException {
         return hasOrExceedsVersion(1, 9);
     }
 
-    public boolean supportCniPortMapping() throws IOException {
+    public boolean supportsCNINetworking() throws IOException {
         // CNI port-mapping is supported by DC/OS 1.10 upwards
-        return hasOrExceedsVersion(1, 10);
+        return hasOrExceedsVersion(1, 9);
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/Capabilities.java]: Is 1.9 the correct value? (I don't know either way). If so should probably update the comment..

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java [hunk 7]
```diff
@@ -18,4 +31,20 @@
...
+            case "dcos":
+                supportsPortMapping = false;
+                break;
+            default:
+                supportsPortMapping = DEFAULT_PORT_MAPPING_POLICY;
+        }
+        return  supportsPortMapping;
+    }
+
+    public static boolean isSupportedNetwork(String networkName) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java]: doesn't look like this is used

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java [hunk 8]
```diff
@@ -10,6 +14,15 @@
     public static final String MESOS_MASTER_ZK_CONNECTION_STRING = MASTER_MESOS + ":2181";
     public static final String MESOS_MASTER_URI = "http://" + MASTER_MESOS;
     public static final Boolean DEFAULT_GPU_POLICY = true;
+    public static final String DEFAULT_OVERLAY_NETWORK = "dcos";
+    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings
+    public static final Set<String> SUPPORTED_OVERLAY_NETWORKS = new HashSet<>(Arrays.asList(DEFAULT_OVERLAY_NETWORK));
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java]: Looks like this isn't used given that isSupportedNetwork isn't called

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java [hunk 9]
```diff
@@ -18,4 +31,20 @@
      * @see com.mesosphere.sdk.state.DefaultSchemaVersionStore
      */
     public static final String SERVICE_ROOT_PATH_PREFIX = "/dcos-service-";
+
+    public static boolean networkSupportsPortMapping(String networkName) {
+        boolean supportsPortMapping;
+        switch (networkName) {
+            case "dcos":
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java]: Could reference `DEFAULT_OVERLAY_NETWORK` here

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java [hunk 10]
```diff
@@ -612,29 +615,24 @@ private static boolean isPortResource(Protos.Resource resource) {
...
-        if (!networkSpec.getPortMappings().isEmpty()) {
+        if (!networkSpec.getPortMappings().isEmpty() &&
+                DcosConstants.networkSupportsPortMapping(networkSpec.getName())) {
+            // we double check the availability of port mapping here in case the service spec was made in
+            // pure java instead of YAML
             for (Map.Entry<Integer, Integer> e : networkSpec.getPortMappings().entrySet()) {
                 Integer hostPort = e.getKey();
                 Integer containerPort = e.getValue();
+                LOGGER.info("Mapping container port {} to host port {}", containerPort, hostPort);
                 netInfoBuilder.addPortMappings(Protos.NetworkInfo.PortMapping.newBuilder()
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java]: Could do:

```
netInfoBuilder.addPortMappingsBuilder()
    .setHostPort(hostPort)
    ...
```

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java [hunk 11]
```diff
@@ -612,29 +615,24 @@ private static boolean isPortResource(Protos.Resource resource) {
...
-
-        if (!networkSpec.getIpAddresses().isEmpty()) {
-            for (String ipAddressString : networkSpec.getIpAddresses()) {
-                netInfoBuilder.addIpAddresses(
-                        Protos.NetworkInfo.IPAddress.newBuilder()
-                                .setIpAddress(ipAddressString)
-                                .setProtocol(Protos.NetworkInfo.Protocol.IPv4)
-                                .build());
-            }
+        if (!networkSpec.getLabels().isEmpty()) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java]: Omit labels from spec for now?

File: sdk/scheduler/src/test/resources/valid-minimal-overlay.yml [hunk 12]
```diff
@@ -28,4 +28,4 @@ pods:
             dest: conf/config-one.conf
           config-two:
             template: config-two.xml.mustache
-            dest: ../other/conf/config-two.xml
+            dest: ../other/conf/config-two.xml
```
Comment (nickbp) [inline sdk/scheduler/src/test/resources/valid-minimal-overlay.yml]: missing a trailing newline

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/Capabilities.java (hunk 6 above)
Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/Capabilities.java]: This is tricky, because I was testing on DC/OS 1.9, so I can't require 1.10..

Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/Capabilities.java]: Could add an override to `SchedulerFlags`? E.g. an envvar like `FORCE_CNI_SUPPORT=[true|false]` that when provided forces support to be enabled or disabled regardless of the detected DC/OS version

Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/Capabilities.java]: actually, I think overlay networks _were_ in 1.9

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java (hunk 7 above)
Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java]: I removed it, then forgot to add it back. It's now used when parsing the YAML and when building the `networkInfo`s.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java (hunk 11 above)
Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java]: This is coming very soon and is already plumbed in the rest of the interfaces. If you feel strongly that we shouldn't have any partially implemented code, I can remove it from `[raw]NetworkSpec` and `DefaultNetworkSpec`.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java (hunk 8 above)
Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/DcosConstants.java]: See below.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java [hunk 13]
```diff
@@ -612,29 +615,29 @@ private static boolean isPortResource(Protos.Resource resource) {
         Protos.NetworkInfo.Builder netInfoBuilder = Protos.NetworkInfo.newBuilder();
         netInfoBuilder.setName(networkSpec.getName());
 
-        if (!networkSpec.getPortMappings().isEmpty()) {
+        if (!DcosConstants.isSupportedNetwork(networkSpec.getName())) {
+            LOGGER.warn(String.format("Virtual netwwork %s is not currently supported, you " +
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java]: `netwwork`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java (hunk 11 above)
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java]: Not a big deal, if the rest is coming soon thats fine.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/YAMLToInternalMappers.java [hunk 14]
```diff
@@ -161,6 +166,52 @@ private static PodSpec from(
...
+            }
+
+            WriteOnceLinkedHashMap<String, RawNetwork> rawNetworks = containerInfoProvider.getNetworks();
+            final Collection<NetworkSpec> networks = new ArrayList<>();
+            if (MapUtils.isNotEmpty(rawNetworks)) {
+                networks.addAll(rawNetworks.entrySet().stream()
+                        .map(rawNetworkEntry -> {
+                            String networkName = rawNetworkEntry.getKey();
+                            if (!DcosConstants.isSupportedNetwork(networkName)) {
+                                LOGGER.warn(String.format("Virtual netwwork %s is not currently supported, you " +
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/YAMLToInternalMappers.java]: `netwwork`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            // versa)
+            PodSpec oldPod = oldPods.get(kv.getKey());
+            if (oldPod != null && podSpecUsesHostPorts(oldPod) != podSpecUsesHostPorts(newPod)) {
+                errors.add(ConfigValidationError.transitionError(
+                        String.format("PodSpec[name:%s]", oldPod.getType()),
+                        oldPod.getNetworks().toString(),
+                        newPod.getNetworks().toString(),
+                        String.format("New config has pod %s moving networks from %s to %s, changing it's " +
+                                        "host ports requirements from %s to %s, not allowed.",
+                                newPod.getType(), oldPod.getNetworks().toString(), newPod.getNetworks().toString(),
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegime.java]: For future reference, you can omit the `.toString()`:
`If the argument is null, then the result is "null". If the argument implements Formattable, then its formatTo method is invoked. Otherwise, the result is obtained by invoking the argument's toString() method.`

Also note: this protects you against throwing NPEs in your logging statements (ask me how I know)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/Capabilities.java (hunk 6 above)
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/Capabilities.java]: I believe testing/master has started reporting its version as 1.10. Confirming with Josh.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java [hunk 16]
```diff
@@ -291,6 +292,8 @@ public OfferRequirement getExistingOfferRequirement(PodInstanceRequirement podIn
             taskInfoBuilder.setDiscovery(getDiscoveryInfo(taskSpec.getDiscovery().get(), podInstance.getIndex()));
         }
 
+        taskInfoBuilder.setContainer(Protos.ContainerInfo.newBuilder().setType(Protos.ContainerInfo.Type.MESOS));
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java]: Does this change mean we will now _always_ be launching in a container?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultNetworkSpec.java [hunk 17]
```diff
@@ -89,14 +77,18 @@ public int hashCode() {
         return HashCodeBuilder.reflectionHashCode(this);
     }
 
+    @Override
+    public String toString() {
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultNetworkSpec.java]: Might as well capture all the fields.

See other classes toString to see how we typically do it (Reflection)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/RawNetwork.java [hunk 18]
```diff
@@ -5,30 +5,28 @@
 
 import java.util.Collections;
 import java.util.List;
+import java.util.Map;
 
 /**
  * Raw YAML network. Null class for now, since only network name is supported, but will gain fields in the future.
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/RawNetwork.java]: Out of date now. TBH I'd just shorten this to `Raw YAML network.`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/YAMLToInternalMappers.java [hunk 19]
```diff
@@ -474,6 +486,15 @@ private static DefaultNetworkSpec from(
         return ports;
     }
 
+    private static boolean maybeUsePortResources(Collection<String> networkNames) {
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/YAMLToInternalMappers.java]: Nit pick: Maybe sounds like it's going to return an optional.

I'd change it to `willUsePortResources`

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegimeTest.java [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+    private PodSpec bridgepod1;
+    @Mock
+    private PodSpec bridgepod2;
+
+    @Mock
+    private PodSpec hostpod1;
+    @Mock
+    private PodSpec hostpod2;
+
+    //private List<PodSpec> overlayPods = new ArrayList<>(Arrays.asList(overlaypod1, overlaypod2));
```
Comment (benclarkwood) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegimeTest.java]: Left overs?

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegimeTest.java [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+        when(hostpod2.getType()).thenReturn(TestConstants.POD_TYPE + "-2");
+        when(hostpod1.getNetworks()).thenReturn(onHostNetwork);
+        when(hostpod2.getNetworks()).thenReturn(onHostNetwork);
+        hostPods.addAll(Arrays.asList(hostpod1, hostpod2));
+
+
+    }
+
+    private void testConfigTransition(List<PodSpec> oldPodSpecs, List<PodSpec> newPodSpecs,
+                                 int expectedHomoTransitionErrors, int expectedHeteroTransitionErrors)
```
Comment (benclarkwood) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegimeTest.java]: nitpick: `expectedSameNetworkErrors` and `expectedChangedNetworkErrors` makes more obvious sense to me.

File: docs/pages/yaml-reference.md [hunk 22]
```diff
@@ -97,6 +93,20 @@ This documentation effectively reflects the Java object tree under [RawServiceSp
...
     This field isn't used! TODO(nickbp) remove.
+   
+  * `networks`
+
+    Allows the pod to join any number of virtual networks on the DC/OS cluster, however the only supported virtual network at present is the `dcos` overlay network. To have the pod join the overlay add the following:
+    ```
+    networks:
+      dcos:
+    ```
+    **N.B.** when a pod joins the overlay it has the following effects:
```
Comment (ArtRand) [inline docs/pages/yaml-reference.md]: https://en.wikipedia.org/wiki/Nota_bene

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java (hunk 16 above)
Comment (ArtRand) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java]: We do already. Adding this proto record only has the effect of telling `navstar` that this is a `MESOS` type container. It's the magic that makes the whole thing work :).

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegimeTest.java (hunk 21 above)
Comment (ArtRand) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/config/validate/PodSpecsCannotChangeNetworkRegimeTest.java]: That's what those words mean...

**Comment by ArtRand** [comment]
retest this please

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/YAMLToInternalMappers.java (hunk 19 above)
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/YAMLToInternalMappers.java]: Nevermind, we use Maybe elsewhere

**Comment by ArtRand** [comment]
retest this please
